Sprint2/step 2

i0ji opened the pull request:
Done so far:
Added 4 states, state transferred to RTK;
it is possible to drag ingredients from the list into the constructor;
it is possible to swap ingredients;

What else is to come:
Get rid of many linter and IDE errors, examine many typing errors, check overall performance

Commit:
03.02.24/b:sprint2/step-2/v:0.1.7.0
refactoring and preparing to redux

Commit:
03.02.24/b:sprint2/step-2/v:0.1.7.0
refactoring and preparing to redux.
All functions and components are set;

Commit:
03.02.24/b:sprint2/step-2/v:0.1.7.0
refactoring and preparing to redux.
All functions and components are set;

i0ji committed:
06.02.24/v:0.1.6.3/b:sprint2/step-2
order details type edit;svg pointer changed

Commit:
07.02.24/v:0.1.6.4/b:sprint2/step-2
DND lib added;OrderDetails type fixed

Commit:
07.02.24/v:0.1.6.4/b:sprint2/step-2
DND lib added;OrderDetails type fixed

i0ji committed:
Merge remote-tracking branch 'origin/sprint2/step-2' into sprint2/step-2

i0ji committed:
08.02.24/v:0.1.6.4/b:sprint2/step-2/
yarn updated;package.json updated

i0ji committed:
refactoring to use RTK

i0ji committed:
RTK started

Commit:
fetch func written and stored via RTK

i0ji committed:
refactoring =(

i0ji committed:
[tag]

Commit:
yarn updated to 4.0.2

Commit:
09.02.24/v:0.1.6.4/sprint2/step-2
rollback to start with RTK;
done first second global state of Ingredient Section;

Commit:
09.02.24/v:0.1.6.4/sprint2/step-2
slowly making second state with many rollbacks

Commit:
10.02.24/v:0.1.6.4/b:sprint2/step-2
second state in progress

Commit:
v:0.1.6.5/b:sprint2/step-2
draggable logic with some artifacts and issues added
"THROUGH TERNII MAT` EGO K ZVEZDAM"

Commit:
setting replace logic

Commit:
functionality for adding and removing elements has been implemented

Commit:
total amount calculating logic set;
adding buns instead of buns?

Commit:
removed not working replace logic

Commit:
removed not working replace logic

Commit:
refactoring

Commit:
11.02.24/v:0.1.6.5/b:sprint2/step-2
last refactoring of constructorSlice and BurgerConstructor

Commit:
ingredients added by type;

Commit:
11.02.24/v:0.1.6.6/b:sprint2/step-2
state logic set:
adding and removing ingredients + updating state;
state of order, total amount of order and current ingredient in modal set;

Commit:
Block and scroll widths are set depending on the number of ingredients

Commit:
small edits; ingredients replace set

Commit:
relocataing ingredeints to single component;

Commit:
relocataing ingredeints to single component;

Commit:
slight progress

Commit:
The drag and drop mechanism is implemented similar with CodeSandbox

Commit:
The drag and drop mechanism is implemented similar with CodeSandbox

Commit:
The drag and drop mechanism is implemented similar with CodeSandbox

Commit:
added "index" to props nad removed an "undefined" error, but still doesnt work

Commit:
reorder logic done and set;
state is working;many IDE and linter mistakes;

Commit:
reorder logic done and set;
state is working;many IDE and linter mistakes;

mikebars1995 commented on src/utils/routs.ts:
```diff
@@ -0,0 +1,3 @@
+// REQUEST URLS
+export const INGREDIENTS_DATA_URL = 'https://norma.nomoreparties.space/api/ingredients';
```
**Нужно исправить**

`'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

mikebars1995 commented on src/utils/burger-api.ts:
```diff
@@ -1,14 +1,8 @@
-            return res.json();
-        })
+    return fetch(INGREDIENTS_DATA_URL)
+        .then(checkResponse)
```
**Можно лучше**

Можно сделать **универсальную функцию запроса с проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

mikebars1995 commented on src/components/BurgerConstructor/BurgerConstructor.tsx:
```diff
@@ -1,118 +1,135 @@
+    const {ingredients: ingredientsData} = useSelector((state: {
+        ingredients: { ingredients: IIngredient[] };
+    }) => state.ingredients);
+    const {isVisible, orderNumber, openModal, closeModal} = useModal(() => createOrder(ingredientsData));
```
**Нужно исправить**


Все **асинхронные** функции проекта нужно сделать в виде `thunk`-функций редакса и объявить их в папке `actions`

mikebars1995 commented on src/components/BurgerIngredients/IngredientGroup/IngredientGroup.tsx:
```diff
@@ -1,44 +1,67 @@
+        const [{isDragging}, drag] = useDrag({
+            type: 'ingredient',
+            item: {
+                id: uuidv4(),
```
**Отлично**


Отлично, что указываете **уникальный** `uuid` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`

mikebars1995 reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на *гитхабе* все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная


Что сделано отлично:

* Отлично, что указываете **уникальный** `uuid` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`

Что нужно исправить:

* По заданию нужно сделать логику **переключения табов**, когда идет **скролл** ингредиентов в списке слева. Это сделать можно с помощью библиотеки `react-intersection-observer`
* При добавлении ингридента в конструктор в списке должна появляться цифра с кол-вом добавленного ингридиента. Скрин https://disk.yandex.ru/i/LaUxx0wBdfhE-A
* `'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.
* Все **асинхронные** функции проекта нужно сделать в виде `thunk`-функций редакса и объявить их в папке `actions`

Что можно улучшить:


* Желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше не давать нажать на кнопку `Оформить заказ`, пока пользователь не добавил булку в заказ.
* лучше отображать **лоадер** (загрузку) при ожидании ответа от сервера с номером заказа
* Можно сделать **универсальную функцию запроса с проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе
* Лучше не использовать пока что `typeScript` и сделать типизацию `propTypes`, в следующих спринтах Вы научитесь правильной типизации `typeScript`


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

**Комментарии `Можно лучше` не обязательны к исправлению прямо сейчас, это рекомендации**

Удачного рефакторинга кода.

i0ji committed:
12.02.24/v0.1.6.6/b:sprint2/step-2
error correction;
added ingredients counter; small css edits to prevent "jumping"; css edits of contructor ingredients list;

i0ji committed:
added bun counter

i0ji committed:
added and refactored URL routs

i0ji committed:
added and refactored URL routs;
started async-thunk functions refactoring;
centered order button;get rid of old API requests;

Commit:
got IDs array in orderSlice

Commit:
got IDs array in orderSlice

Commit:
got IDs array in orderSlice

i0ji committed:
13.02.24/v:0.1.7.0/b:sprint2/step-2
small state issues resolved;
still trying to use useModal and orderSlice

i0ji committed:
no ideas...

i0ji committed:
no ideas...

i0ji committed:
v:0.1.7.0.1
changing the css to expand the total visible area, better interaction

i0ji committed:
v:0.1.7.0.1
changing the css to expand the total visible area, better interaction

i0ji committed:
v:0.1.7.0.1
changing the css to expand the total visible area, better interaction

i0ji committed:
scrollIntoView attempts; immutability-helper removed; right _id of ingredients selected

Commit:
Merge remote-tracking branch 'origin/sprint2/step-2' into sprint2/step-2

# Conflicts:
#	src/components/BurgerConstructor/BurgerConstructor.tsx
#	src/components/BurgerIngredients/IngredientGroup/IngredientGroup.tsx
#	src/components/Modal/OrderDetails/OrderDetails.tsx
#	src/hooks/useModal.ts
#	src/services/slices/orderSlice.ts

Commit:
get order number logic set;

Commit:
scroll to view from scratch

Commit:
The scrolling logic is implemented with minor flaws.

i0ji committed:
updated readme with a summary of all commits

mikebars1995 commented on src/services/slices/orderSlice.ts:
```diff
@@ -0,0 +1,53 @@
+		},
+		body: JSON.stringify(requestBody),
+	});
+	const data = await response.json();
```
**Нужно исправить**


Любой ответ от сервера нужно проверять на `ok` (на **корректность**) и вызывать ошибку, если ответ не `ok`. Таким образом, в `catch` Вы будете ловить и ошибки сервера, и ошибки невалидных данных.

mikebars1995 reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Остались некоторые недочеты: 

* Любой ответ от сервера нужно проверять на `ok` (на **корректность**) и вызывать ошибку, если ответ не `ok`. Таким образом, в `catch` Вы будете ловить и ошибки сервера, и ошибки невалидных данных.



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

i0ji committed:
README updated;
reset ingredients list after getting orderNumber;;
some type issues fixed;
added checkResponse to orderNumber request;
adjusting scroll accuracy

i0ji committed:
14.02.24/b:sprint3/step-1/v:0.1.7.1:
minor edits; OIA scroll calibration;

i0ji committed:
preloader and error message added;

i0ji committed:
v: 0.1.7.1: preloader and error message added;

i0ji committed:
readme update

i0ji committed:
readme update/modal typed issue fixed

i0ji committed:
[tag]

Commit:
v:0.1.7.1.2;
layout and markup changes: buttons hover/favico/global positioning/ingredients hovers

Commit:
v:0.1.7.1.2;
Initial bun attempts

mikebars1995 reviewed (commented):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.

Commit:
15.02.24/b:sprint2/step-2/v:0.1.7.2
Initial bun set; стилизация кода, приведение в порядок

Commit:
15.02.24/b:sprint2/step-2/v:0.1.7.2
Initial bun set; стилизация кода, приведение в порядок

i0ji committed:
16.02.24/b:sprint2/step-2/v:0.1.7.2
displaying the initial of buns;
